packages/next/src/client/components/router-reducer/reducers/server-action-reducer.ts

```diff
@@ -111,27 +115,24 @@ async function fetchServerAction(
   const contentType = res.headers.get('content-type')
 
   if (contentType === RSC_CONTENT_TYPE_HEADER) {
-    const response: ActionFlightResponse = await createFromFetch(
-      Promise.resolve(res),
-      {
-        callServer,
-      }
-    )
+    const response: RSCPayload = await createFromFetch(Promise.resolve(res), {
+      callServer,
+    })
+
+    const actionFlightData = getFlightDataFromRSCPayload(response)
+    const actionResult = getActionResultFromRSCPayload(response)
 
     if (location) {
       // if it was a redirection, then result is just a regular RSC payload
-      const [, actionFlightData] = (response as any) ?? []
       return {
-        actionFlightData: actionFlightData,
+        actionFlightData,
         redirectLocation,
         revalidatedParts,
       }
     }
 
-    // otherwise it's a tuple of [actionResult, actionFlightData]
-    const [actionResult, [, actionFlightData]] = (response as any) ?? []
     return {
-      actionResult,
+      actionResult: actionResult ?? undefined,
```

Comment:
Fallback wouldn't be needed if the types were a bit stricter, and/or we wouldn't be overly cautious with casting known server responses, see above.

Reply:
I set out to make [these types](https://github.com/vercel/next.js/blob/canary/packages/next/src/client/components/router-reducer/reducers/server-action-reducer.ts#L46-L47) non-optional so it's more explicit when they aren't provided, but I wanted to try and keep unrelated refactors to a minimum